// rc-scroll-anim/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scrollScreen = exports.Event = exports.Link = exports.Element = exports.Parallax = exports.OverPack = undefined;

var _ScrollOverPack = require('./ScrollOverPack');

var _ScrollOverPack2 = _interopRequireDefault(_ScrollOverPack);

var _ScrollParallax = require('./ScrollParallax');

var _ScrollParallax2 = _interopRequireDefault(_ScrollParallax);

var _ScrollLink = require('./ScrollLink');

var _ScrollLink2 = _interopRequireDefault(_ScrollLink);

var _ScrollElement = require('./ScrollElement');

var _ScrollElement2 = _interopRequireDefault(_ScrollElement);

var _EventDispatcher = require('./EventDispatcher');

var _EventDispatcher2 = _interopRequireDefault(_EventDispatcher);

var _ScrollScreen = require('./ScrollScreen');

var _ScrollScreen2 = _interopRequireDefault(_ScrollScreen);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// export this package's api
var OverPack = exports.OverPack = _ScrollOverPack2['default'];
var Parallax = exports.Parallax = _ScrollParallax2['default'];
var Element = exports.Element = _ScrollElement2['default'];
var Link = exports.Link = _ScrollLink2['default'];
var Event = exports.Event = _EventDispatcher2['default'];
var scrollScreen = exports.scrollScreen = _ScrollScreen2['default'];

exports['default'] = {
  OverPack: _ScrollOverPack2['default'],
  Parallax: _ScrollParallax2['default'],
  Element: _ScrollElement2['default'],
  Link: _ScrollLink2['default'],
  Event: _EventDispatcher2['default'],
  scrollScreen: _ScrollScreen2['default']
};
